### PR TITLE
Make sure Teleport sessions use the user login state.

### PR DESCRIPTION
**Pull request by mdwn**
The user login state is being generated properly but not being used for the various sessions. UserLoginState is now properly integrated into sessions. Additionally, it was noticed that user login state was not being used for Github logins, so it's been added there as well.

The consequence of this is that, though `tsh` logins work as expected, UI logins (and potentially others) don't take on the roles given by the access list. By adding the user state into the session flow, we get the expected roles in the UI.

Additionally, getUserOrLoginState has been exposed for use by the enterprise repo.

### CODE REVIEW

File: lib/auth/github.go [hunk 1]
```diff
@@ -622,6 +622,11 @@ func (a *Server) validateGithubAuthCallback(ctx context.Context, diagCtx *SSODia
 		return nil, trace.Wrap(err)
 	}
 
+	userState, err := a.GetUserOrLoginState(ctx, user.GetName())
```
Comment (camscale) [inline lib/auth/github.go:625]: Do you intend to make this same change for SAML and OIDC in the enterprise repo? It looks like `validateOIDCAuthCallback()` and `validateSAMLResponse()` also create the web session with values from a `User` resource and not a `UserState`.

**Review by camscale — COMMENTED** [review]
What were the consequences of this bug? Can you add that to the PR description? It's not entirely clear to me why this change is needed and what it actually does.

File: lib/auth/github.go (hunk 1 above)
Comment (mdwn) [reply lib/auth/github.go:625]: I do, that's actually the reason I made this `GetUserOrLoginState` public here.

**Comment by mdwn** [comment]
> What were the consequences of this bug? Can you add that to the PR description? It's not entirely clear to me why this change is needed and what it actually does.

Clarified!

**Review by jakule — APPROVED** [review]

**Review by kimlisa — APPROVED** [review]

**Pull request merged**